doc/man3/X509_cmp.pod

```diff
@@ -25,16 +25,20 @@ This set of functions are used to compare X509 objects, including X509
 certificates, X509 CRL objects and various values in an X509 certificate.
 
 The X509_cmp() function compares two B<X509> objects indicated by parameters
-B<a> and B<b>. The comparison is based on the B<memcmp> result of the hash
+I<a> and I<b>. The comparison is based on the B<memcmp> result of the hash
 values of two B<X509> objects and the canonical (DER) encoding values.
 
 The X509_NAME_cmp() function compares two B<X509_NAME> objects indicated by
-parameters B<a> and B<b>. The comparison is based on the B<memcmp> result of
-the canonical (DER) encoding values of the two objects. L<i2d_X509_NAME(3)>
-has a more detailed description of the DER encoding of the B<X509_NAME> structure.
+parameters I<a> and I<b>. The comparison is based on the B<memcmp> result of the
+canonical (DER) encoding values of the two objects using L<i2d_X509_NAME(3)>.
+This procedure adheres to the matching rules for Distinguished Names (DN)
+given in RFC 4517 section 4.2.15 and RFC 5280 section 7.1.
+In particular, the order of Relative Distinguished Names (RDNs) is relevant.
+On the other hand, if an RDN is multi-valued, i.e., it contains a set of
+AttributeValueAssertions (AVAs), its members are effectively not ordered.
```

Comment:
I find this description rather confusing. RFC5280 section 7.1 has MUSTs that can not be done using memcmp(), including case folding, and insignificant space handling.

It's also confusing that you say the AVAs are not ordered, so one has to wonder why a memcmp() should work, unless you know that the DER encoding requires ordering them by tag.

Reply:
Initially I had the same concerns,
but the dumb `memcpy()` is called after normalizing the DN strings, which apparently does fulfill the MUSTs of the RFC. In particular this sorts the AVAs in multi-valued RDNs, which makes them order independent, as I found by experimenting with `X509_name_cmp()`.

Reply:
So how about just saying that it's normalized before doing the memcmp() in the manpage?

Reply:
Looking at asn1_string_canon(), it does not seem to be doing everything required by RFC4518.

Reply:
My focus on that PR was fixes regarding multi-valued RDNs,
so I did not investigate further into the undocumented function `asn1_string_canon()`.
Can well be that it does not do everything required by the RFC(s).

Reply:
How about the following extended/clarified documentation of `X509_NAME_cmp()`?
```
The X509_NAME_cmp() function compares two B<X509_NAME> objects B<a> and B<b>.
The two Distinguished Name objects are compared using B<memcmp>
after normalizing them (using the internal function B<asn1_string_canon()>)
encoding them using DER.
L<i2d_X509_NAME(3)> has a more detailed description of the DER encoding
of B<X509_NAME> structures.
```

Reply:
I really don't like implementation details in the manual. I suggest something like:
The X509_NAME_cmp() function compares two B<X509_NAME> objects B<a> and B<b>. The comparison is done after conversion to the canonical form.

Reply:
Good point that the external documentation should be implementation agnostic.
Yet I would prefer the term "normalized" over "canonical" because this is the one used in RFC 5280.
It would also be good to give a little more detail on which conversions are done;
at least it should be made clear which of the required normalization steps are (not) done.